src/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,371 @@
+    /// in the Z-Y'-X" order (where Z -> yaw, Y -> pitch, X -> roll). 
+    /// The return value is a list of [roll, pitch, yaw]
+    pub fn to_euler_angles(&self) -> [f64;3] {
+        // get a normalized version of the quaternion
+        let quat = self.get_normalized();
+
+        // calculate yaw
+        let yaw_sin_pitch_cos = 2.0 * ((quat.real * quat.k) + (quat.i * quat.j));
+        let yaw_cos_pitch_cos = 1.0 - 2.0 * ((quat.j * quat.j) + (quat.k * quat.k));
+        let yaw = yaw_sin_pitch_cos.atan2(yaw_cos_pitch_cos);
+
+        // calculate pitch and roll
+        let pitch_sin = 2.0 * ((quat.real * quat.j) - (quat.k * quat.i));
+        let pitch: f64;
+        let roll: f64;
+        // for a pitch that is π / 2, we experience gimbal lock
+        // and must calculate roll based on the real rotation and yaw
+        if pitch_sin.abs() >= 1.0 {
+            pitch = (std::f64::consts::PI / 2.0).copysign(pitch_sin);
+            roll = (2.0 * quat.i.atan2(quat.real)) + yaw.copysign(pitch_sin);
+        } else {
+            pitch = pitch_sin.asin();
+            let roll_sin_pitch_cos = 2.0 * ((quat.real * quat.i) + (quat.j * quat.k));
+            let roll_cos_pitch_cos = 1.0 - 2.0 * ((quat.i * quat.i) + (quat.j * quat.j));
+            roll = roll_sin_pitch_cos.atan2(roll_cos_pitch_cos);
+        }
+
+
+        [roll, pitch, yaw]
+    }
+
+    /// Return the imaginary components of a quaternion as
+    /// a 3-vector
+    pub fn imag(&self) -> Vector3 {
+        Vector3 { x: self.i, y: self.j, z: self.k }
+    }
+
+    /// Sets the imaginary components of a quaternion from
+    /// a provided 3-vector
+    pub fn set_imag_from_vector(&mut self, imag: Vector3) {
+        self.i = imag.x;
+        self.j = imag.y;
+        self.k = imag.z;
+    }
+
+    /// Returns whether the quaternion is normalized 
+    /// (a four-vector on the unit sphere in quaternion space)
+    pub fn is_normalized(&self) -> bool {
+        approx_eq!(f64, self.norm2().sqrt(), 1.0)
+    }
+
+    pub fn norm2(&self) -> f64 {
+        (self.real * self.real) + (self.i * self.i) + (self.j * self.j) + (self.k * self.k)
+    }
+
+    /// Normalizes a quaternion
+    pub fn normalize(&mut self) {
+        self.scale(1.0 / self.norm2().sqrt())
+    }
+
+    /// Returns a normalized copy of the quaternion
+    pub fn get_normalized(&self) -> Self {
+        let mut copy_quat = self.clone();
+        copy_quat.normalize();
+        copy_quat
+    }
+
+    pub fn scale(&mut self, factor: f64) {
+        self.real = self.real * factor;
+        self.i = self.i * factor;
+        self.j = self.j * factor;
+        self.k = self.k * factor;
+    }
+
+    pub fn get_scaled(&self, factor: f64) -> Self {
+        let mut copy_quat = self.clone();
+        copy_quat.scale(factor);
+        copy_quat
+    }
+
+    pub fn conjugate(&self) -> Self {
+        Self { real: self.real, i: self.i * -1.0, j: self.j * -1.0, k: self.k * -1.0 }
```

Comment:
DIscussion question: I had expected `-1.0 * self.i`, rather than `self.i * -1.0`, because normally when you write math, the numbers come before the letters. Do other folks prefer this ordering instead? I'm not against it, but it did surprise me a little.